### PR TITLE
Draft for flow

### PR DESCRIPTION
**Pull request by Keats**

### CODE REVIEW

**Comment by Keats** [comment]
@teh I should you should co-write this one

File: content/articles/flow-and-typescript-1.md [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+For some people it means foregoing JavaScript itself and using languages such as [Clojurescript](https://github.com/clojure/clojurescript), [Elm](http://elm-lang.org/) or [Purescript](http://www.purescript.org/) while others want to keep using JavaScript, especially now that ES6 makes writing it tolerable.
+
+A simple way to make a codebase easier to maintain and to grow is to have types and that is exactly what [Flow](https://babeljs.io/) and [TypeScript](http://www.typescriptlang.org/) offer, albeit in different ways.
+<!-- PELICAN_END_SUMMARY -->
+
+TypeScript is a ES6 transpiler like [Babel](https://babeljs.io/) with types while Flow only does the annotation and static type checking. If you want to use ES6 with Flow, you will need to use Babel as well.
+
+For this series of articles, I am going to convert my [react-example](https://github.com/Keats/react-example) which is a very basic react/redux/react-router Kanban board app to both Flow and TypeScript. Both TypeScript and Flow versions are in the same [repo](https://github.com/Keats/flow-typescript) but only the Flow one is working at the time of writing.
```
Comment (Gautier) [inline content/articles/flow-and-typescript-1.md]: I would capitalise "React/Redux/react-router"

File: content/articles/flow-and-typescript-1.md [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+
+In this part 1, we will look at Flow.
+
+## Installation and usage
+
+It's important to say that Flow doesn't currently work on Windows. You can have a look at [issue #6](https://github.com/facebook/flow/issues/6) on their tracker if you are using Windows. Running it in a [docker image](https://hub.docker.com/r/motiz88/flow/) might work for you though.
+
+For the sake of simplicity I will install the Flow binary with [flow-bin](https://www.npmjs.com/package/flow-bin).instead of using e.g. a gulp plugin (mainly since I couldn't find one).
```
Comment (Gautier) [inline content/articles/flow-and-typescript-1.md]: Is this an extra `.` between the link and instead?

File: content/articles/flow-and-typescript-1.md [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+## An example
+
+Let's start with the [`routes.js`](https://github.com/Keats/flow-typescript/blob/master/flow/src/app/routes.js) files and we get the following error:
+
+```
+  5: import { Route, IndexRoute } from "react-router";
+                                       ^^^^^^^^^^^^^^ react-router. Required module not found
+
+```
+As you can see, Flow doesn't complain about the React import as it comes bundled with an interface file for it.
```
Comment (Gautier) [inline content/articles/flow-and-typescript-1.md]: I think it's not obvious that `routes.js` starts with `import React from "react";`, it took me a minute to re-read this

File: content/articles/flow-and-typescript-1.md [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+
+An interface file is how Flow knows what to expect from an external import, such as react-router.
+
+To fix our error, we need to add an interface file for react-router. The interface file will be empty for now and we tell flow to use the `typings` directory for libraries:
```
Comment (Gautier) [inline content/articles/flow-and-typescript-1.md]: we tell Flow (capitalisation)

File: content/articles/flow-and-typescript-1.md [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+
+Flow doesn't currently have enums (sum types) but allows some approximation with [disjoint union types](http://flowtype.org/blog/2015/07/03/Disjoint-Unions.html):
+
+```js
+type TodoAction = Add | Remove
+type Add = { type: 'Add', entry: string }
+type Remove = { type: 'Remove', id: number }
+```
+
+With this setup flow will complain if 1) the type doesn't match any of the defined types or 2) the structure doesn't match. E.g.
```
Comment (Gautier) [inline content/articles/flow-and-typescript-1.md]: With this setup Flow <- capitalisation (unless you were mentioning the binary file itself?)

File: content/articles/flow-and-typescript-1.md [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+will fail with a similar error because the `entry` field is missing:
+
+```
+ 20: let add: TodoAction = {type: 'Add'}
+                           ^^^^^^^^^^^^^ object literal. This type is incompatible with
+ 20: let add: TodoAction = {type: 'Add'}
+              ^^^^^^^^^^ union type
+```
+
+Unfortunately this is no true sum type. E.g. the following example will work just fine even though `"omgbug"` is not valid for the `type` field:
```
Comment (Gautier) [inline content/articles/flow-and-typescript-1.md]: "Unfortunately this is no true sum type." <- I think this is correct but a bit unusual "This is no moon!", is there any word missing?

File: content/articles/flow-and-typescript-1.md [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  switch (action.type) {
+    case "omgbug":
+      return "bug";
+  }
+}
+```
+
+## The state of Flow
+
+Flow type inference is really good and the opt-in aspect makes it really easy to add it to a codebase.
```
Comment (Gautier) [inline content/articles/flow-and-typescript-1.md]: 2 "really"s in the same sentence. What about "Flow type inference is useful/good/... and the opt-in aspect makes it really easy to add it to a codebase."?

File: content/articles/flow-and-typescript-1.md [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+- Non-existent support in tools such as Webpack.
+- Low priority project for Facebook from an outsider point of view: ImmutableJS (another Facebook project) for example does not have an official Flow interface file.
+- The documentation is somewhat incomplete
+- Flow runs a server in the background which needs to be restarted manually, e.g. when adding a new interface file. Forgetting this will lead to some head-scratching error messages.
+- Flow is not super compatible with Babel and es6-lint. E.g. `function returns_func(): function` works in Flow but not in Babel because `function` is not a valid annotation in Babel (syntax error).
+
+Now that Flow supports ES6, it's mostly lacking in tooling and documentation.
+
+Writing interface files myself is OK (well maybe not all of [Lodash](https://lodash.com/docs) in one go) but it would be nice to have a good reference on how to write proper interface files, and to have Facebook provide interface files for their own projects.
```
Comment (Gautier) [inline content/articles/flow-and-typescript-1.md]: myself <- I assume this post is written as T+V

File: content/articles/flow-and-typescript-1.md [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+The only official files we found are part of flow itself: [https://github.com/facebook/flow/blob/master/lib/react.js](https://github.com/facebook/flow/blob/ed8f3d136d3432651fd39544d7ca40244a7423c2/lib/react.js) and hasn't been updated yet to 0.14.
```
Comment (Gautier) [inline content/articles/flow-and-typescript-1.md]: are part of Flow (capitalisation)

**Pull request merged**